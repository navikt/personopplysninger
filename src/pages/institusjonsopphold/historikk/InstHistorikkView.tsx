import { useState } from 'react';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import { FormattedMessage } from 'react-intl';
import { Link, useLocation } from 'react-router-dom';
import { useIntlFormatter } from 'hooks/useIntlFormatter';
import { InstInfo } from 'types/inst';
import Kilde from 'components/kilde/Kilde';
import { CustomHelpText } from 'components/customHelpText/CustomHelpText';
import { Alert, BodyShort, Button, Label } from '@navikt/ds-react';
import { ChevronDownIcon } from '@navikt/aksel-icons';
import dayjs from 'dayjs';

const InstHistorikkView = (props: { instInfo: InstInfo }) => {
    const [viewAmount, setViewAmount] = useState(20);
    const location = useLocation();
    const { formatMessage } = useIntlFormatter();
    const { instInfo } = props;

    let animateDelay = 0;
    let animateDelayKey = 0;
    let animateDelaySum = 0;

    return (
        <>
            <Alert variant="info">
                <FormattedMessage
                    id="inst.disclaimer"
                    values={{
                        lenkeTilInstMelding: (text) => <Link to={'https://www.nav.no/samarbeidspartner/opphold-institusjon'}>{text}</Link>,
                    }}
                />
            </Alert>
            <div className={'inst__tabell'}>
                {instInfo.length > 0 ? (
                    <>
                        <div className="historikk__flex-rad inst__head">
                            <div className="historikk__flex-kolonne">
                                <Label as="p">
                                    <FormattedMessage id="inst.periode" />
                                </Label>
                            </div>
                            <div className="historikk__flex-kolonne">
                                <Label as="p">
                                    <FormattedMessage id="inst.institusjon" />
                                </Label>
                            </div>
                        </div>
                        <TransitionGroup>
                            {instInfo
                                .sort((a, b) => (dayjs(a.startdato) > dayjs(b.startdato) ? -1 : 1))
                                .slice(0, viewAmount)
                                .map((innslag, i) => {
                                    if (animateDelayKey >= 20) {
                                        animateDelay = 0;
                                        animateDelayKey = 0;
                                        animateDelaySum = 0;
                                    }
                                    animateDelay = 50 + animateDelayKey * 15;
                                    animateDelaySum = +animateDelay;
                                    animateDelayKey++;

                                    const startdato = dayjs(innslag.startdato).format('DD.MM.YYYY');
                                    const faktiskSluttdato = innslag.faktiskSluttdato ? dayjs(innslag.faktiskSluttdato).format('DD.MM.YYYY') : '';

                                    return (
                                        <CSSTransition
                                            key={i}
                                            classNames={'inst__animate'}
                                            style={{
                                                transitionDelay: `${animateDelay}ms`,
                                                fontWeight: i >= 20 && i >= viewAmount - 20 ? 'bold' : 'normal',
                                            }}
                                            timeout={100 + animateDelaySum}
                                        >
                                            <div className="historikk__flex-rad">
                                                <div className="historikk__flex-kolonne historikk__heading">
                                                    <BodyShort>{`${startdato} - ${faktiskSluttdato}`}</BodyShort>
                                                    {innslag.fiktivSluttdato && (
                                                        <CustomHelpText title={formatMessage('inst.fiktivSluttdato.tittel')}>
                                                            <FormattedMessage id={'inst.fiktivSluttdato'} />
                                                        </CustomHelpText>
                                                    )}
                                                </div>
                                                <div className="historikk__flex-kolonne">
                                                    <Link to={`${location.pathname}/${innslag.registreringstidspunkt}`} className="lenke">
                                                        <BodyShort>{innslag.institusjonsnavn}</BodyShort>
                                                    </Link>
                                                </div>
                                            </div>
                                        </CSSTransition>
                                    );
                                })}
                        </TransitionGroup>
                    </>
                ) : (
                    <div className="historikk__ingen-data">
                        <Alert variant="info">
                            <FormattedMessage id="inst.ingendata" />
                        </Alert>
                    </div>
                )}
                {instInfo.length > 20 && instInfo.length >= viewAmount && (
                    <div className={'inst__se-flere'}>
                        <Button
                            variant={'tertiary'}
                            onClick={() => setViewAmount(viewAmount + 20)}
                            icon={<ChevronDownIcon aria-hidden={true} />}
                            iconPosition="right"
                        >
                            Se flere
                        </Button>
                    </div>
                )}
            </div>
            <div className="inst__kilde">
                <Kilde kilde="inst.kilde" lenkeType="INGEN" />
            </div>
        </>
    );
};

export default InstHistorikkView;
