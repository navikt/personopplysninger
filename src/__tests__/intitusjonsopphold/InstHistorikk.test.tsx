import '@testing-library/jest-dom';
import { IntlProvider } from 'react-intl';
import nbMessages from 'text/nb';
import InstHistorikkView from '../../pages/institusjonsopphold/historikk/InstHistorikkView';
import instInfo from '../../clients/apiMock/app/fetch/inst-info.json';
import { render } from '@testing-library/react';

jest.mock('react-router-dom', () => ({
    Link: () => <div />,
    useLocation: () => ({
        pathname: 'pathname',
    }),
}));

describe('InstHistorikkView', () => {
    it('renders correctly', () => {
        const { asFragment } = render(
            <IntlProvider locale={'nb'} messages={nbMessages}>
                <InstHistorikkView instInfo={instInfo} />
            </IntlProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
