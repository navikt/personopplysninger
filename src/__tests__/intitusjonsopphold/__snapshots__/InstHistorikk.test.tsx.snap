// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstHistorikkView renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arbeidsforhold__disclaimer"
  >
    <div
      class="navds-alert navds-alert--info navds-alert--medium"
    >
      <svg
        aria-labelledby="title-r0"
        class="navds-alert__icon"
        fill="none"
        focusable="false"
        height="1em"
        role="img"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title
          id="title-r0"
        >
          Informasjon
        </title>
        <path
          clip-rule="evenodd"
          d="M12 0c6.627 0 12 5.373 12 12s-5.373 12-12 12S0 18.627 0 12 5.373 0 12 0ZM9 19v-2h2v-5H9v-2h4v7h2v2H9Zm3-14a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
      <div
        class="navds-alert__wrapper navds-body-long"
      >
        <p
          class="navds-body-long"
        >
          Her kan du se opplysninger om opphold på alders- og sykehjem, helseinstitusjoner og fengsel fra 1.januar 2017. Oppdager du feil, må du kontakte institusjonen direkte.
        </p>
      </div>
    </div>
    <div
      class="inst__tabell"
    >
      <div
        class="historikk__flex-rad inst__head"
      >
        <div
          class="historikk__flex-kolonne"
        >
          <p
            class="navds-label"
          >
            Periode
          </p>
        </div>
        <div
          class="historikk__flex-kolonne"
        >
          <p
            class="navds-label"
          >
            Institusjon
          </p>
        </div>
      </div>
      <div>
        <div
          class="historikk__flex-rad"
          style="transition-delay: 50ms; font-weight: normal;"
        >
          <div
            class="historikk__flex-kolonne historikk__heading"
          >
            <p
              class="navds-body-short"
            >
              29.12.2017 - 17.12.2018
            </p>
          </div>
          <div
            class="historikk__flex-kolonne"
          >
            <div />
          </div>
        </div>
        <div
          class="historikk__flex-rad"
          style="transition-delay: 65ms; font-weight: normal;"
        >
          <div
            class="historikk__flex-kolonne historikk__heading"
          >
            <p
              class="navds-body-short"
            >
              29.12.2017 - 20.12.2020
            </p>
            <div
              class="customHelpText"
            >
              <button
                class="customHelpText__help-button navds-button navds-button--tertiary navds-button--medium"
                type="button"
              >
                <span
                  aria-live="polite"
                  class="navds-label"
                >
                  <svg
                    class="customHelpText__icon"
                    fill="none"
                    focusable="false"
                    height="1em"
                    role="img"
                    viewBox="0 0 24 24"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M12 0c6.627 0 12 5.373 12 12s-5.373 12-12 12S0 18.627 0 12 5.373 0 12 0Zm0 2C6.477 2 2 6.477 2 12s4.477 10 10 10 10-4.477 10-10S17.523 2 12 2Zm0 14a1.5 1.5 0 1 1 0 3 1.5 1.5 0 0 1 0-3Zm0-11c2.205 0 4 1.657 4 3.693 0 .986-.416 1.914-1.172 2.612l-.593.54-.294.28c-.477.466-.869.94-.936 1.417l-.01.144v.814h-1.991v-.814c0-1.254.84-2.214 1.675-3.002l.74-.68c.38-.35.59-.816.59-1.31 0-1.024-.901-1.856-2.01-1.856-1.054 0-1.922.755-2.002 1.71l-.006.145H8C8 6.657 9.794 5 12 5Z"
                      fill="currentColor"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </button>
              <div
                aria-hidden="true"
                class="navds-popover customHelpText__popover navds-popover--hidden"
                data-placement="top"
                style="position: absolute; top: 0px; left: 0px;"
                tabindex="-1"
              >
                <div
                  class="navds-popover__content"
                >
                  Sluttdato for oppholdet er ikke mottatt. NAV har avsluttet oppholdet maskinelt.
                </div>
                <div
                  class="navds-popover__arrow"
                  style="bottom: -0.5rem;"
                />
              </div>
            </div>
          </div>
          <div
            class="historikk__flex-kolonne"
          >
            <div />
          </div>
        </div>
        <div
          class="historikk__flex-rad"
          style="transition-delay: 80ms; font-weight: normal;"
        >
          <div
            class="historikk__flex-kolonne historikk__heading"
          >
            <p
              class="navds-body-short"
            >
              13.06.2017 - 15.09.2017
            </p>
          </div>
          <div
            class="historikk__flex-kolonne"
          >
            <div />
          </div>
        </div>
      </div>
    </div>
    <div
      class="inst__kilde"
    >
      <div
        class="kilde__container"
      >
        <div
          class="kilde__tekst"
        >
          <p
            class="navds-detail"
          >
            Kilde: Den enkelte institusjon
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
